### PR TITLE
Fix errors when submitting empty feature or feature values ppv2

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.1.x
| Description?      | fixes https://github.com/PrestaShop/PrestaShop/issues/31401, https://github.com/PrestaShop/PrestaShop/issues/31402
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | https://github.com/PrestaShop/PrestaShop/issues/31401, https://github.com/PrestaShop/PrestaShop/issues/31402
| Fixed ticket?     | https://github.com/PrestaShop/PrestaShop/issues/31401, https://github.com/PrestaShop/PrestaShop/issues/31402
| Related PRs       | If theme, autoupgrade or other module change is needed to make this change work, provide a link to related PRs here.
| Sponsor company   | Your company or customer's name goes here (if applicable).

### CODE REVIEW

File: src/PrestaShopBundle/Form/Admin/Sell/Product/EventListener/FeatureValueListener.php [hunk 1]
```diff
@@ -90,14 +104,22 @@ public function updateFeatureValuesOptions(FormEvent $event): void
...
             ],
-        ]);
+            'constraints' => [],
+        ];
+        if (!$hasCustomValue) {
+            $options['constraints'][] = new NotBlank([
+                'message' => $this->translator->trans('Choose a value or provide a customized one', [], 'Admin.Catalog.Feature'),
+            ]);
+        }
+        $newFeatureValueForm = $this->formCloner->cloneForm($form->get('feature_value_id'), $options);
```
Comment (zuk3975) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/EventListener/FeatureValueListener.php]: this might seem abit strange, but idk how else to solve it. (been working around for some time and this whole setup seems to work while nothing else did)
So the issue was:

placeholder can only contain NULL value. 
In js when we reset fields, we cannot make it NULL (its either empty string, '0', 0, but it is not equivalent to null anyway). I suspect that was the root problem.
So there was different behavior depending if it was submitted after js reset or not
(1. if after page load you don't change the FeatureId, but only set Feature value to "choose value" option, then you'll get error ` Feature value missing, specify a feature_value_id or new custom_values` - this is solved by adding NotBlank constraint.
2. if you change featureId option, then js loads the feature values and then if you don't select any feature value, but leave "choose value" option, you'll get error complaining about invalid option). - this validation already worked, I just altered the option invalid_message

So basically what I've done is manage 2 different errors so that they look the same (they are in theory the same errors from UX perspective, but technically in symfony context they are understood differently)

**Review by jolelievre — COMMENTED** [review]
Thanks @zuk3975 

you just need to update FeatureValueListenerTest.php according to your changes in the constructor (and maybe improve the tests themselves to test the new behaviour you added)

File: src/PrestaShopBundle/Form/Admin/Sell/Product/Details/FeatureValueType.php [hunk 2]
```diff
@@ -71,20 +72,27 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+                        'message' => $this->trans('Choose a feature', 'Admin.Catalog.Feature'),
+                    ]),
+                ],
             ])
             ->add('feature_value_id', ChoiceType::class, [
                 'required' => false,
                 'empty_data' => null,
                 'placeholder' => $this->trans('Choose a value', 'Admin.Catalog.Feature'),
                 'label' => $this->trans('Pre-defined value', 'Admin.Catalog.Feature'),
+                'invalid_message' => $this->trans('Choose a value or provide a customized one', 'Admin.Catalog.Feature'),
```
Comment (zuk3975) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/Details/FeatureValueType.php:94]: ping @l-delin for wording.

**Review by matthieu-rolland — DISMISSED** [review]
lgtm :+1:

**Review by jolelievre — DISMISSED** [review]
Thanks @zuk3975

**Review by florine2623 — COMMENTED** [review]
Hello @zuk3975 ,

Link to automated tests : https://github.com/florine2623/testing_pr/actions/runs/4263064185

 #31401 - No exception, just a warning : ✅ 
Screenshot 2023-02-24 at 16 48 03

But when I continue to fill my product page,
- I choose Feature : Property > I Save
- I have an exception (Same as in  #31402)
`An unexpected error occurred. [PrestaShop\PrestaShop\Core\Domain\Product\FeatureValue\Exception\InvalidProductFeatureValuesFormatException code 0]: Feature value missing, specify a feature_value_id or new custom_values`

https://user-images.githubusercontent.com/16019289/221222525-c2df8a0a-1ff7-42a1-aafd-f20c09a08a02.mov


#31402 is not fixed yet, as you can see from the screen record :

https://user-images.githubusercontent.com/16019289/221224928-f78e4052-f3f1-4626-8a98-7c063c0b6ef4.mov

The exception is the same as previously : 
`An unexpected error occurred. [PrestaShop\PrestaShop\Core\Domain\Product\FeatureValue\Exception\InvalidProductFeatureValuesFormatException code 0]: Feature value missing, specify a feature_value_id or new custom_values`

Could you check please ? ^^
Thanks!

**Comment by zuk3975** [comment]
thanks @florine2623, indeed I can reproduce. 
It's very strange - form fires constraint error but still reaches command and fires that InvalidProductFeatureValuesFormatException :cold_sweat: 
image.
I'll try to fix it

File: src/PrestaShopBundle/Form/Admin/Sell/Product/Details/FeatureValueType.php (hunk 2 above)
Comment (l-delin) [reply src/PrestaShopBundle/Form/Admin/Sell/Product/Details/FeatureValueType.php:94]: Hello, looks good to me :-)

**Comment by zuk3975** [comment]
So apparently NotBlank constraint wasn't working as expected and the reason invalid choice wasn't working all the time was in related to missing "disabled" attr at some point.
Seems to work with latest changes (the issue is quite tricky though, so hopefully now its ok :sweat: )

File: src/PrestaShopBundle/Form/Admin/Sell/Product/EventListener/FeatureValueListener.php [hunk 3]
```diff
@@ -93,7 +83,7 @@ public function updateFeatureValuesOptions(FormEvent $event): void
 
         if (empty($data['feature_id'])) {
             // make sure feature_value_id form is disabled in case it was rendered after error of not selecting feature_id
-            $newFeatureValueForm = $this->formCloner->cloneForm($form->get('feature_value_id'), ['disabled' => true]);
```
Comment (zuk3975) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/EventListener/FeatureValueListener.php]: this seemed to cause troubles. It must have been "attr" => disabled" or else it would still be disabled and avoid validation or something like that (not sure, it was quite tricky to find this :smile: )

File: admin-dev/themes/new-theme/js/pages/product/edit/manager/feature-values-manager.ts [hunk 4]
```diff
@@ -108,48 +108,50 @@ export default class FeatureValuesManager {
       const $featureValueSelector = $(ProductMap.featureValues.featureValueSelect, $collectionRow).first();
       $featureValueSelector.prop('disabled', hasCustomValue);
       if (hasCustomValue) {
-        $featureValueSelector.val('');
+        $featureValueSelector.val('0');
```
Comment (zuk3975) [inline admin-dev/themes/new-theme/js/pages/product/edit/manager/feature-values-manager.ts:111]: this way when its submitted it triggers "invalid option" error, or else I think it is considered as a placeholder value and doesn't trigger that error (though also not sure as its quite tricky :smile: )

**Review by jolelievre — DISMISSED** [review]
Thanks @zuk3975 

Not sure I understood all the subtility 😅 But I trust you and QA to validate this behaviour, the code seems legit anyway

**Review by matks — DISMISSED** [review]

**Review by MhiriFaten — COMMENTED** [review]
Hello @zuk3975 ,

I have checked your PR and I noticed those issues : 

**issue 1 :** 

1. Go to BO > Advanced parameters > New & Experimental features > Enable New product page
2. Go to BO > Catalog > Products > Create a product V2
3. Click on Add a feature
4. Set these values :

    > **Feature:** Choose a feature | **Pre-defined value :** Choose a feature | **OR Customized value :** empty
    > **Feature:** Composition | **Pre-defined value :** Choose a feature | **OR Customized value :** empty
    > **Feature:** Property  | **Pre-defined value :** Choose a feature | **OR Customized value :** any value (exp:12)
    > 

5. Click on Save
6. See error of Details ==> This value is not valid
7. Click again on Pre-defined value and choose again Choose a value 
8. Change the OR Customized value to any other value
9. Click on Save
10. See error :warning:  The warning of the second feature is disappeared !

[Untitled_ Mar 2, 2023 11_55 AM.webm](https://user-images.githubusercontent.com/111756615/222411998-0748524e-3dfa-420f-b703-34a585d991ac.webm)

**Issue 2 :**

1. Set the first feature :

    > **Feature:** Choose a feature | **Pre-defined value :** Choose a feature | **OR Customized value :** empty

2. Click on Save
3. See error :warning:  The exception is displayed !

[Untitled_ Mar 2, 2023 12_14 PM.webm](https://user-images.githubusercontent.com/111756615/222414605-6e0768ff-1aaa-42aa-a838-74e52d1e62ad.webm)

Could you recheck it please ?
Thank you :pray:

**Comment by zuk3975** [comment]
oh gosh, this feature is killing me, Im out of options how to fix everything. Fixing one case breaks other 100 cases and there seems to be endless count of different scenarios :sweat:

**Comment by zuk3975** [comment]
Fixed. One important option "empty_data = 0" was missing.
I've tested your provided scenarios, it seems to work as expected now :pray:

File: tests/Integration/PrestaShopBundle/Form/EventListener/FeatureValueListenerTest.php [hunk 5]
```diff
@@ -313,13 +323,16 @@ class SimpleFeaturesFormTest extends CommonAbstractType
...
+            // these are the default options which are important and might be affected by listener
             'disabled' => true,
             'attr' => [
                 'disabled' => true,
                 'data-toggle' => 'select2',
                 'class' => 'feature-value-selector',
             ],
+            // empty data = 0 is important. Without it being set to 0, the "invalid option" constraint is not triggered,
+            // and the form is not invalidated (which should happen because field is not optional therefore 0 is invalid option)
+            'empty_data' => 0,
```
Comment (zuk3975) [inline tests/Integration/PrestaShopBundle/Form/EventListener/FeatureValueListenerTest.php]: this is the important part. It allows invalidating placeholder value.
Without this, the form is considered valid (because the default null value is understood as optiona/placeholder value by symfony). With this, once placeholder is submitted it is converted to 0, which causes invalid option error as expected.

**Review by jolelievre — DISMISSED** [review]
Thanks @zuk3975 

Let's hope it's the last one 💪

**Review by 0x346e3730 — DISMISSED** [review]

**Review by MhiriFaten — COMMENTED** [review]
Hello @zuk3975 ,

I tried to retest the PR but as you see I couldn't install it

image

Could you check it please?
Thank you !

**Comment by zuk3975** [comment]
@MhiriFaten it seems to be your localhost issues, not related to PR.
Anyway I need to rebase the PR

File: src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml [hunk 6]
```diff
@@ -1840,6 +1840,7 @@ services:
     arguments:
       - '@prestashop.adapter.form.choice_provider.feature_values_choice_provider'
       - '@PrestaShopBundle\Form\FormCloner'
+      - '@translator'
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml]: Seems like it's an error (maybe related to the rebase) but the listener doesn't use translator in its constructor

**Review by jolelievre — CHANGES_REQUESTED** [review]

**Review by jolelievre — DISMISSED** [review]

**Review by kpodemski — DISMISSED** [review]

**Review by florine2623 — COMMENTED** [review]
Hello @zuk3975 ,

Link to automated tests : https://github.com/florine2623/testing_pr/actions/runs/4365454384

https://github.com/PrestaShop/PrestaShop/issues/31401 is fixed ✅ 

https://user-images.githubusercontent.com/16019289/223750654-1bcc6a04-a2b1-4612-8dec-7ce6f9bebd5d.mov

---

[Issues 1 and 2](https://github.com/PrestaShop/PrestaShop/pull/31508#pullrequestreview-1321632274) from @MhiriFaten are fixed ✅ 

---

By testing https://github.com/PrestaShop/PrestaShop/issues/31402, I have found another exception 😓 

Screenshot 2023-03-08 at 16 27 33

https://user-images.githubusercontent.com/16019289/223756012-69d7c011-ee81-46d9-92d5-623fc8a20e79.mov

`An unexpected error occurred. [PrestaShop\PrestaShop\Core\Domain\Feature\Exception\FeatureValueConstraintException code 20]: Invalid FeatureValue localized property "value" for language with id "1"`

Here are the steps to reproduce :
1. Add a feature > Only add a customized value > Save and publish OK ✅ 
2. Choose a feature (ie Composition) > Add a customized value > Save and publish OK ✅ 
3. Delete your customized value > Save and publish NOK :x:

Could you check ? 
Thanks a lot 🤗

**Review by jolelievre — APPROVED** [review]
Thanks @zuk3975

File: src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml [hunk 7]
```diff
@@ -1479,13 +1479,13 @@ services:
...
+  PrestaShopBundle\Form\Admin\Sell\Product\Details\FeatureValueType:
     parent: 'form.type.translatable.aware'
+    autowire: true
+    public: false
     arguments:
-      - '@prestashop.adapter.form.choice_provider.features_choice_provider'
-      - '@form.type.sell.product.event_listener.feature_value_listener'
-    public: true
+      $featuresChoiceProvider: '@prestashop.adapter.form.choice_provider.features_choice_provider'
+      $featureValueListener: '@PrestaShopBundle\Form\Admin\Sell\Product\EventListener\FeatureValueListener'
```
Comment (lartist) [inline src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml:1488]: Why not using auto wiring ?

File: src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml [hunk 8]
```diff
@@ -1835,12 +1835,11 @@ services:
...
-  form.type.sell.product.event_listener.feature_value_listener:
-    class: 'PrestaShopBundle\Form\Admin\Sell\Product\EventListener\FeatureValueListener'
+  PrestaShopBundle\Form\Admin\Sell\Product\EventListener\FeatureValueListener:
+    public: false
+    autowire: true
     arguments:
-      - '@prestashop.adapter.form.choice_provider.feature_values_choice_provider'
-      - '@PrestaShopBundle\Form\FormCloner'
-    public: true
+      $featureValuesChoiceProvider: '@prestashop.adapter.form.choice_provider.feature_values_choice_provider'
```
Comment (lartist) [inline src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml:1842]: Why not using autowiring?

File: src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml (hunk 7 above)
Comment (zuk3975) [reply src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml:1488]: it depends on interface, so we must define which actual class to inject

File: src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml (hunk 8 above)
Comment (zuk3975) [reply src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml:1842]: it depends on interface, so we must define which actual class to inject

**Review by lartist — APPROVED** [review]

**Review by djoelleuch — APPROVED** [review]
Hello @zuk3975 

Automated tests : OK (https://github.com/djoelleuch/ga.tests.ui.pr/actions/runs/4382692293)
 No exception, just a warning :heavy_check_mark: 

[Untitled_ Mar 10, 2023 11_13 AM.webm](https://user-images.githubusercontent.com/115152460/224290535-36d3b84a-edd5-412f-abd3-b38b4847f661.webm)


LGTM ! It is QA approved 
Thank you :rocket:

**Comment by FabienPapet** [comment]
Thank you @zuk3975

**Comment by futurowebpl** [comment]
Hello, 

This is not working when New product page in Experimental functions is enabled...

**Comment by kpodemski** [comment]
@futurowebpl 

this fix is specifically for a new product page, so if it doesn’t work then please report a bug
